crates/ra_hir/src/debug.rs

```diff
@@ -0,0 +1,83 @@
+//! printf debugging infrastructure for rust-analyzer.
+//!
+//! When you print a hir type, like a module, using `eprintln!("{:?}", module)`,
+//! you usually get back a numeric ID, which doesn't tell you much:
+//! `Module(92)`.
+//!
+//! This module adds convenience `debug` methods to various types, which resolve
+//! the id to a human-readable location info:
+//!
+//! ```not_rust
+//! eprintln!("{:?}", module.debug(db));
+//! =>
+//! Module { name: collections, path: "liballoc/collections/mod.rs" }
+//! ```
+//!
+//! Note that to get this info, we might need to execute queries! So
+//!
+//! * don't use the `debug` methods for logging
+//! * when debugging, be aware that interference is possible.
+
+use std::fmt;
+
+use ra_db::{CrateId, FileId};
+
+use crate::{db::HirDatabase, Crate, Module, Name};
+
+impl Crate {
+    pub fn debug(self, db: &impl HirDebugDatabase) -> impl fmt::Debug + '_ {
+        debug_fn(move |fmt| db.debug_crate(self, fmt))
+    }
+}
+
+impl Module {
+    pub fn debug(self, db: &impl HirDebugDatabase) -> impl fmt::Debug + '_ {
+        debug_fn(move |fmt| db.debug_module(self, fmt))
+    }
+}
+
+pub trait HirDebugHelper: HirDatabase {
+    fn crate_name(&self, _krate: CrateId) -> Option<String> {
+        None
+    }
+    fn file_path(&self, _file_id: FileId) -> Option<String> {
+        None
+    }
+}
+
+pub trait HirDebugDatabase {
+    fn debug_crate(&self, krate: Crate, fmt: &mut fmt::Formatter<'_>) -> fmt::Result;
+    fn debug_module(&self, module: Module, fmt: &mut fmt::Formatter<'_>) -> fmt::Result;
+}
+
+impl<DB: HirDebugHelper> HirDebugDatabase for DB {
+    fn debug_crate(&self, krate: Crate, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut builder = fmt.debug_tuple("Crate");
+        match self.crate_name(krate.crate_id) {
+            Some(name) => builder.field(&name),
```

Comment:
Maybe show crate id in any case. That would prevent confusion with multiple versions of a crate.

Reply:
I prefer to print less by default, as "too verbose logging" is harder to fix in a specific case. It's possible to print the `Crate` itself to see the ID